# Question 1


In [21]:
import pandas as pd

# URL of the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# Load the dataset into a Pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print("First few rows of the dataset:")
print(df.head())

# Check for missing values in each column
missing_values = df.isnull().sum()

# Print the columns with missing values
print("\nMissing values in each column:")
print(missing_values)


First few rows of the dataset:
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            3734

# Question 2

1. The Titanic dataset has 891 rows and 12 columns.

2. An observation is an individual characteristic (or a row). For example, in the Titanic dataset, each observation represents a different person. A variable is a characteristic for each observation (or column). In our dataset, they are things like age or fare. 

# Question 3


In [5]:
# Numeric columns summary
numeric_summary = df.describe()
print(numeric_summary)

# Categorical columns summary
categorical_summary = df.describe(include=['object'])
print(categorical_summary)


       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  
                           Name   Sex  Ticket    Cabin Embarked
count            

# Question 4

In [6]:
df.shape

(891, 12)

In [7]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

We only have 7 out of 12 columns so 5 columns worth of information is missing. We only have the ages of 714 out of 891 people. This is because df.describe only gives information for numeric variables. 

In [8]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Here we have all the columns, including non-numerical ones.


# Question 5.



An attribute is a characteristic of an object and holds its value. Does not require parentheses. 
A method is a function that gets associated to an object. Requires parentheses.

ChatGPT summary:

In this session, I worked with the Titanic dataset in a pandas DataFrame. We began by determining the shape of the dataset, revealing that it contains 891 rows (observations) and 12 columns (variables). I learned about the definitions of observations (rows, representing individual passengers) and variables (columns, representing attributes such as age, fare, and survival).

Next, I was provided with code to generate summaries for numeric and categorical columns using df.describe() for deeper analysis. I also learned the difference between attributes and methods:

Attributes (like df.shape) are properties of the DataFrame that provide information without needing parentheses.
Methods (like df.describe()) perform operations or calculations and require parentheses to be executed.
Finally, I explored practical examples of these concepts using pandas methods and attributes, with code that can be directly applied to analyze the dataset.

https://chatgpt.com/share/d10a7587-0be7-4b22-8dd8-b53c5cf202be

# Question 6



Count is the number of non-missing values in each variable Mean is the average of values in a variable, 
Std or standard deviation is the spread of the values in a variable (closer or farther to the mean),
Min is the minimum (lowest) value in the variable,
25% is the first quartile, a quarter of the data points lie below it,
50% is the median, half the data points are above it and half are below,
75% is the third quartile, 75% of the data points lie below is,
Max is the maximum (largest) value of the variable.

# Question 7


1. When we want a dataset without any missing values and don't want to delete any rows.
2. When we want to simply remove a row from the dataset without changing anything else. 
3. No matter which one we apply first, the 'col' column will be deleted, but if we apply df.dropna() beforehand even the observations with only the 'col' values missing will be deleted. But if we delete the 'col' column first, the observations with only the 'col' value missing will remain as they will no longer have any values missing

In [9]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [10]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Since a large majority of observations are missing their cabin value, we should remove the column to keep all the information. There is also quite an important amount of observations missing their age value, but not enough (I don't think) to remove the whole column, as it is also important information. Finally, two people are also missing their Embarked value which is also not enough to delete the column.

In [11]:
del df["Cabin"]

In [12]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
dtype: int64

In [13]:
df.dropna()

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
885      5            382652  29.1250        Q  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[712 rows x 11 columns]

We are down from 891 rows and 12 columns to 712 rows and 11 columns

Chatbot summary:

Here’s a summary of the conversation since the last summary:

df.describe() Method:

We discussed the statistical summary provided by the df.describe() method. It generates important statistics like:
Count: Number of non-null values.
Mean: Arithmetic average of values.
Std (Standard Deviation): Measure of data spread.
Min: Minimum value in the variable.
25%, 50% (Median), 75%: Quartile values that give insight into the distribution.
Max: Maximum value.
KeyError Issue:

You encountered a KeyError when attempting to delete the 'Cabin' column. This error occurs when the specified column does not exist in the DataFrame.
Suggested solutions include checking for typos in the column name, printing the DataFrame’s columns to confirm existence, and safely deleting the column by using a conditional statement to check if it exists.
This summary covers the key points discussed since the last update.

https://chatgpt.com/share/d10a7587-0be7-4b22-8dd8-b53c5cf202be

New chatbot session staring here

# Question 8
1

Let's restore the dataset

In [27]:
df = pd.read_csv(url)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

df.groupby("col1")["col2"].describe() groups the dataframe by the "col1" values and gives you the statistics for "col2". So if we were to use Pclass as "col1" and Survived as "col2", we would find the statistics of survival by class in the Titanic. 

In [33]:
df.groupby("Pclass")["Survived"].describe()

count      mean       std  min  25%  50%  75%  max
Pclass                                                    
1       216.0  0.629630  0.484026  0.0  0.0  1.0  1.0  1.0
2       184.0  0.472826  0.500623  0.0  0.0  0.0  1.0  1.0
3       491.0  0.242363  0.428949  0.0  0.0  0.0  0.0  1.0

2

In [16]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [23]:
df.groupby("Pclass")["Survived"].describe()

count      mean       std  min  25%  50%  75%  max
Pclass                                                    
1       216.0  0.629630  0.484026  0.0  0.0  1.0  1.0  1.0
2       184.0  0.472826  0.500623  0.0  0.0  0.0  1.0  1.0
3       491.0  0.242363  0.428949  0.0  0.0  0.0  0.0  1.0

We get different values in the two count columns because they are not measuring the same thing. df.describe is counting every non-missing value for each column, while df.groupby("Pclass")["Survived"].describe() counts the number of people in each passenger class.

3

A. Found the issue immediatly 

B. Found the issue immediatly 

C. Found the issue immediatly 

D. Found the issue immediatly 

E. Found the issue immediatly 

F. Found the issue immediatly 

G. Found the issue immediatly

# Question 9

I did use those resources

ChatGPT summary

Here's a summary of our chat:

Initial Question:

You asked about the function of df.groupby("col1")["col2"].describe(). I explained that it groups data by col1 and gives descriptive statistics (like mean, min, max) for col2 within each group.
Errors in Code:

You encountered several errors while trying to work with the Titanic dataset.
NameError: The variable df was not defined, so I suggested loading the dataset into a DataFrame.
HTTPError: There was an issue with loading the dataset due to an incorrect URL, which I corrected.
NameError (DF not defined): This occurred due to using DF instead of df (case-sensitive).
SyntaxError: There was an unclosed parenthesis in the pd.read_csv() function, which I helped fix.
AttributeError: You used pd.readcsv() instead of pd.read_csv().
KeyError ('pclass'): You attempted to use "pclass" instead of the correct column name "Pclass".
NameError (Survived not defined): This happened because you used Survived without quotes, and I corrected it to ["Survived"].
Final Fix:

I guided you to use the correct column names and provided proper syntax for grouping and describing the data.
Let me know if you need further clarification!

https://chatgpt.com/share/66e3a326-6e98-8013-8df4-c7d61ba09b19